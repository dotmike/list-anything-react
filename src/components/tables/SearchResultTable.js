import React, { Component } from "react";
import PropTypes from "prop-types";
import { Table } from "semantic-ui-react";
import sortBy from "lodash.sortby";

class SearchResultTable extends Component {
  constructor(props) {
    super(props);

    this.state = {
      column: null,
      data: props.results,
      direction: null
    };
  }

  componentWillReceiveProps = nextProps => {
    if (this.props.results !== nextProps.results) {
      this.setState({ data: nextProps.results });
    }
  };

  handleSort = clickedColumn => () => {
    const { column, data, direction } = this.state;

    if (column !== clickedColumn) {
      this.setState({
        column: clickedColumn,
        data: sortBy(data, [clickedColumn]),
        direction: "ascending"
      });
    } else {
      this.setState({
        data: data.reverse(),
        direction: direction === "ascending" ? "descending" : "ascending"
      });
    }
  };

  render() {
    const { onTableRowClick } = this.props;
    const { column, data, direction } = this.state;

    return (
      <Table celled compact="very" selectable sortable>
        <Table.Header>
          <Table.Row>
            <Table.HeaderCell
              sorted={column === "value" ? direction : null}
              onClick={this.handleSort("value")}
            >
              Value
            </Table.HeaderCell>
            <Table.HeaderCell
              sorted={column === "tab" ? direction : null}
              onClick={this.handleSort("tab")}
            >
              Tab
            </Table.HeaderCell>
            <Table.HeaderCell
              sorted={column === "subject" ? direction : null}
              onClick={this.handleSort("subject")}
            >
              Subject
            </Table.HeaderCell>
          </Table.Row>
        </Table.Header>
        <Table.Body style={{ cursor: "pointer" }}>
          {!data.length && (
            <Table.Row>
              <Table.Cell width={16}>Nothing was found.</Table.Cell>
            </Table.Row>
          )}
          {data.map(result => (
            <Table.Row
              key={result._id}
              to={`/subject-data/${result._id}`}
              onClick={onTableRowClick}
            >
              <Table.Cell collapsing>{result.value}</Table.Cell>
              <Table.Cell collapsing>{result.tab}</Table.Cell>
              <Table.Cell collapsing>{result.subject}</Table.Cell>
            </Table.Row>
          ))}
        </Table.Body>
      </Table>
    );
  }
}

SearchResultTable.propTypes = {
  // ownProps
  results: PropTypes.arrayOf(
    PropTypes.shape({
      _id: PropTypes.string,
      value: PropTypes.string,
      subject: PropTypes.string,
      tab: PropTypes.string
    })
  ).isRequired,
  onTableRowClick: PropTypes.func.isRequired
};

export default SearchResultTable;
