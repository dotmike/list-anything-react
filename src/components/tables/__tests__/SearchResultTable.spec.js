import SearchResultTable from "../SearchResultTable";

const testData = {
  results: [
    {
      _id: "0",
      value: "Test Value",
      subject: "Test Subject",
      tab: "Test Tab",
      field: {
        _id: "0",
        description: "Test Description",
        field_type: "test_field"
      }
    }
  ]
};

const setup = (propOverrides = {}) => {
  const props = {
    results: testData.results,
    onTableRowClick: jest.fn(),
    ...propOverrides
  };

  return {
    props,
    // Lembrar de sempre chamar `unmount` quando usar esse wrapper \/
    wrapperMount: wrapperMount(SearchResultTable, props)
  };
};

describe("SearchResultTable", () => {
  it("renders correctly when `results` is empty", () => {
    const { wrapperMount: wrapper } = setup({
      results: []
    });
    expect(wrapper()).toMatchSnapshot();

    wrapper().unmount();
  });

  it("renders correctly when `results` is not empty", () => {
    const { wrapperMount: wrapper } = setup();
    expect(wrapper()).toMatchSnapshot();

    wrapper().unmount();
  });

  it("calls `onTableRowClick` when clicking on a `TableBody > TableRow`", () => {
    const { wrapperMount: wrapper, props } = setup();
    expect(wrapper()).toMatchSnapshot();

    wrapper()
      .find("TableBody")
      .find("TableRow")
      .first()
      .simulate("click");

    expect(props.onTableRowClick).toHaveBeenCalled();

    wrapper().unmount();
  });
});
