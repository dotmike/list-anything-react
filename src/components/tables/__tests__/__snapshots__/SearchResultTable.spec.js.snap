// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResultTable calls \`onTableRowClick\` when clicking on a \`TableBody > TableRow\` 1`] = `
<SearchResultTable
  onTableRowClick={[Function]}
  results={
    Array [
      Object {
        "_id": "0",
        "field": Object {
          "_id": "0",
          "description": "Test Description",
          "field_type": "test_field",
        },
        "subject": "Test Subject",
        "tab": "Test Tab",
        "value": "Test Value",
      },
    ]
  }
>
  <Table
    as="table"
    celled={true}
    compact="very"
    selectable={true}
    sortable={true}
  >
    <table
      className="ui celled selectable sortable very compact table"
    >
      <SortableDataTableHeader
        data={
          Array [
            Object {
              "_id": "0",
              "field": Object {
                "_id": "0",
                "description": "Test Description",
                "field_type": "test_field",
              },
              "subject": "Test Subject",
              "tab": "Test Tab",
              "value": "Test Value",
            },
          ]
        }
        renderHeader={[Function]}
        sort={[Function]}
      >
        <TableHeader
          as="thead"
        >
          <thead
            className=""
          >
            <TableRow
              as="tr"
              cellAs="td"
            >
              <tr
                className=""
              >
                <TableHeaderCell
                  as="th"
                  key="value"
                  onClick={[Function]}
                  sorted={null}
                >
                  <TableCell
                    as="th"
                    className=""
                    onClick={[Function]}
                  >
                    <th
                      className=""
                      onClick={[Function]}
                    >
                      Value
                    </th>
                  </TableCell>
                </TableHeaderCell>
                <TableHeaderCell
                  as="th"
                  key="field"
                  onClick={[Function]}
                  sorted={null}
                >
                  <TableCell
                    as="th"
                    className=""
                    onClick={[Function]}
                  >
                    <th
                      className=""
                      onClick={[Function]}
                    >
                      Field
                    </th>
                  </TableCell>
                </TableHeaderCell>
                <TableHeaderCell
                  as="th"
                  key="tab"
                  onClick={[Function]}
                  sorted={null}
                >
                  <TableCell
                    as="th"
                    className=""
                    onClick={[Function]}
                  >
                    <th
                      className=""
                      onClick={[Function]}
                    >
                      Tab
                    </th>
                  </TableCell>
                </TableHeaderCell>
                <TableHeaderCell
                  as="th"
                  key="subject"
                  onClick={[Function]}
                  sorted={null}
                >
                  <TableCell
                    as="th"
                    className=""
                    onClick={[Function]}
                  >
                    <th
                      className=""
                      onClick={[Function]}
                    >
                      Subject
                    </th>
                  </TableCell>
                </TableHeaderCell>
              </tr>
            </TableRow>
          </thead>
        </TableHeader>
      </SortableDataTableHeader>
      <DataTableBody
        renderBody={[Function]}
      >
        <TableBody
          as="tbody"
          style={
            Object {
              "cursor": "pointer",
            }
          }
        >
          <tbody
            className=""
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            <TableRow
              as="tr"
              cellAs="td"
              key="0"
              onClick={[Function]}
              to="/subject-data/0"
            >
              <tr
                className=""
                onClick={[Function]}
                to="/subject-data/0"
              >
                <TableCell
                  as="td"
                >
                  <td
                    className=""
                  >
                    <div
                      key="0"
                    />
                  </td>
                </TableCell>
                <TableCell
                  as="td"
                  collapsing={true}
                >
                  <td
                    className="collapsing"
                  >
                    Test Description
                  </td>
                </TableCell>
                <TableCell
                  as="td"
                  collapsing={true}
                >
                  <td
                    className="collapsing"
                  >
                    Test Tab
                  </td>
                </TableCell>
                <TableCell
                  as="td"
                  collapsing={true}
                >
                  <td
                    className="collapsing"
                  >
                    Test Subject
                  </td>
                </TableCell>
              </tr>
            </TableRow>
          </tbody>
        </TableBody>
      </DataTableBody>
    </table>
  </Table>
</SearchResultTable>
`;

exports[`SearchResultTable renders correctly when \`results\` is empty 1`] = `
<SearchResultTable
  onTableRowClick={[Function]}
  results={Array []}
>
  <Table
    as="table"
    celled={true}
    compact="very"
    selectable={true}
    sortable={true}
  >
    <table
      className="ui celled selectable sortable very compact table"
    >
      <SortableDataTableHeader
        data={Array []}
        renderHeader={[Function]}
        sort={[Function]}
      >
        <TableHeader
          as="thead"
        >
          <thead
            className=""
          >
            <TableRow
              as="tr"
              cellAs="td"
            >
              <tr
                className=""
              >
                <TableHeaderCell
                  as="th"
                  key="value"
                  onClick={[Function]}
                  sorted={null}
                >
                  <TableCell
                    as="th"
                    className=""
                    onClick={[Function]}
                  >
                    <th
                      className=""
                      onClick={[Function]}
                    >
                      Value
                    </th>
                  </TableCell>
                </TableHeaderCell>
                <TableHeaderCell
                  as="th"
                  key="field"
                  onClick={[Function]}
                  sorted={null}
                >
                  <TableCell
                    as="th"
                    className=""
                    onClick={[Function]}
                  >
                    <th
                      className=""
                      onClick={[Function]}
                    >
                      Field
                    </th>
                  </TableCell>
                </TableHeaderCell>
                <TableHeaderCell
                  as="th"
                  key="tab"
                  onClick={[Function]}
                  sorted={null}
                >
                  <TableCell
                    as="th"
                    className=""
                    onClick={[Function]}
                  >
                    <th
                      className=""
                      onClick={[Function]}
                    >
                      Tab
                    </th>
                  </TableCell>
                </TableHeaderCell>
                <TableHeaderCell
                  as="th"
                  key="subject"
                  onClick={[Function]}
                  sorted={null}
                >
                  <TableCell
                    as="th"
                    className=""
                    onClick={[Function]}
                  >
                    <th
                      className=""
                      onClick={[Function]}
                    >
                      Subject
                    </th>
                  </TableCell>
                </TableHeaderCell>
              </tr>
            </TableRow>
          </thead>
        </TableHeader>
      </SortableDataTableHeader>
      <DataTableBody
        renderBody={[Function]}
      >
        <TableBody
          as="tbody"
          style={
            Object {
              "cursor": "pointer",
            }
          }
        >
          <tbody
            className=""
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            <TableRow
              as="tr"
              cellAs="td"
            >
              <tr
                className=""
              >
                <TableCell
                  as="td"
                  width={16}
                >
                  <td
                    className="sixteen wide"
                  >
                    Nothing was found.
                  </td>
                </TableCell>
              </tr>
            </TableRow>
          </tbody>
        </TableBody>
      </DataTableBody>
    </table>
  </Table>
</SearchResultTable>
`;

exports[`SearchResultTable renders correctly when \`results\` is not empty 1`] = `
<SearchResultTable
  onTableRowClick={[Function]}
  results={
    Array [
      Object {
        "_id": "0",
        "field": Object {
          "_id": "0",
          "description": "Test Description",
          "field_type": "test_field",
        },
        "subject": "Test Subject",
        "tab": "Test Tab",
        "value": "Test Value",
      },
    ]
  }
>
  <Table
    as="table"
    celled={true}
    compact="very"
    selectable={true}
    sortable={true}
  >
    <table
      className="ui celled selectable sortable very compact table"
    >
      <SortableDataTableHeader
        data={
          Array [
            Object {
              "_id": "0",
              "field": Object {
                "_id": "0",
                "description": "Test Description",
                "field_type": "test_field",
              },
              "subject": "Test Subject",
              "tab": "Test Tab",
              "value": "Test Value",
            },
          ]
        }
        renderHeader={[Function]}
        sort={[Function]}
      >
        <TableHeader
          as="thead"
        >
          <thead
            className=""
          >
            <TableRow
              as="tr"
              cellAs="td"
            >
              <tr
                className=""
              >
                <TableHeaderCell
                  as="th"
                  key="value"
                  onClick={[Function]}
                  sorted={null}
                >
                  <TableCell
                    as="th"
                    className=""
                    onClick={[Function]}
                  >
                    <th
                      className=""
                      onClick={[Function]}
                    >
                      Value
                    </th>
                  </TableCell>
                </TableHeaderCell>
                <TableHeaderCell
                  as="th"
                  key="field"
                  onClick={[Function]}
                  sorted={null}
                >
                  <TableCell
                    as="th"
                    className=""
                    onClick={[Function]}
                  >
                    <th
                      className=""
                      onClick={[Function]}
                    >
                      Field
                    </th>
                  </TableCell>
                </TableHeaderCell>
                <TableHeaderCell
                  as="th"
                  key="tab"
                  onClick={[Function]}
                  sorted={null}
                >
                  <TableCell
                    as="th"
                    className=""
                    onClick={[Function]}
                  >
                    <th
                      className=""
                      onClick={[Function]}
                    >
                      Tab
                    </th>
                  </TableCell>
                </TableHeaderCell>
                <TableHeaderCell
                  as="th"
                  key="subject"
                  onClick={[Function]}
                  sorted={null}
                >
                  <TableCell
                    as="th"
                    className=""
                    onClick={[Function]}
                  >
                    <th
                      className=""
                      onClick={[Function]}
                    >
                      Subject
                    </th>
                  </TableCell>
                </TableHeaderCell>
              </tr>
            </TableRow>
          </thead>
        </TableHeader>
      </SortableDataTableHeader>
      <DataTableBody
        renderBody={[Function]}
      >
        <TableBody
          as="tbody"
          style={
            Object {
              "cursor": "pointer",
            }
          }
        >
          <tbody
            className=""
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            <TableRow
              as="tr"
              cellAs="td"
              key="0"
              onClick={[Function]}
              to="/subject-data/0"
            >
              <tr
                className=""
                onClick={[Function]}
                to="/subject-data/0"
              >
                <TableCell
                  as="td"
                >
                  <td
                    className=""
                  >
                    <div
                      key="0"
                    />
                  </td>
                </TableCell>
                <TableCell
                  as="td"
                  collapsing={true}
                >
                  <td
                    className="collapsing"
                  >
                    Test Description
                  </td>
                </TableCell>
                <TableCell
                  as="td"
                  collapsing={true}
                >
                  <td
                    className="collapsing"
                  >
                    Test Tab
                  </td>
                </TableCell>
                <TableCell
                  as="td"
                  collapsing={true}
                >
                  <td
                    className="collapsing"
                  >
                    Test Subject
                  </td>
                </TableCell>
              </tr>
            </TableRow>
          </tbody>
        </TableBody>
      </DataTableBody>
    </table>
  </Table>
</SearchResultTable>
`;
